### PR TITLE
Improvements to domain validation

### PR DESCRIPTION
**Pull request by rsa33**
* Prevent the creation of new domains when no DNS records are present for either of `$domain` and `www.$domain`.
  (Domains resolving to a different server are still permitted so that users can migrate to us without downtime.)

* Add better validation for domain document roots: enforce being inside `public_html`, block special characters that may interfere with macros, check the directory already exists, and fix non-canonical paths.

* Apply the docroot checks consistently across both adding and editing domains.

### CODE REVIEW

File: control/templates/shared/add_vhost_test.html [hunk 1]
```diff
@@ -54,7 +54,7 @@ <h2>Test custom domain</h2>
     <input type="hidden" name="confirm" value="on">
     {% if valid[domain] == (true, true) or valid[prefixed] == (true, true) %}
         <input type="submit" class="btn btn-outline-primary" value="Looks good, add domain">
-    {% else %}
+    {% elif valid[domain] != (none, none) or valid[domain] != (none, none) %}
```
Comment (andyhhp) [inline control/templates/shared/add_vhost_test.html]: Do you really mean for both sides of the `or` to be the same, or should one be `valid[prefixed]` ?

File: control/webapp/utils.py [hunk 2]
```diff
@@ -91,6 +92,33 @@ def parse_domain_name(domain):
         return domain.encode("idna").decode("ascii")
 
 
+def validate_domain_docroot(owner, path):
+    if not path:
+        return path, None
+    if any(ch in path for ch in string.whitespace + "\\" + '"' + "'"):
+        return path, "Document roots cannot contain spaces or quotes."
```
Comment (andyhhp) [inline control/webapp/utils.py]: "spaces, backslashes or quotes" ?

File: control/templates/shared/add_vhost_test.html (hunk 1 above)
Comment (rsa33) [reply control/templates/shared/add_vhost_test.html]: Indeed, one should be checking `prefixed` -- ultimately we're looking for at least one valid result (each tuple is a pair of IPv4, IPv6 results) before the user is allowed to add the domain.

File: control/webapp/utils.py [hunk 3]
```diff
@@ -91,6 +92,33 @@ def parse_domain_name(domain):
...
+        path = path.replace("public_html/", "", 1)
+    if isinstance(owner, Member):
+        username = owner.crsid
+        top = "home"
+    else:
+        username = owner.society
+        top = "societies"
+    base = os.path.join("/public", top, username, "public_html")
+    target = os.path.abspath(os.path.join(base, path))
+    if not target.startswith(base):
```
Comment (smbz) [inline control/webapp/utils.py]: I'm worried about the edge case with e.g. `path="../public_html_2"` - should this be a test with os.path.commonpath rather than a string operation?

Comment (rsa33) [reply control/webapp/utils.py]: Good catch -- I may have been imagining `base` ended with a slash here, but using the `os.path` libs is safer in any case.

**Review by andyhhp — APPROVED** [review]
